File: src/main/python/com/example/tutorial/dsa/medium/logicgames/card_shuffler.py
"""
CardShuffler
----------------------------------
This program implements the Fisher-Yates shuffle algorithm for cards.
The core problem solved here is Shuffle an Array (LeetCode #384).

Problem Statement:
    Given an integer array nums, design an algorithm to randomly shuffle the array.
    All permutations should be equally likely.

Real UseCase:
    In a credit card offers system:
    - Randomize offer presentation order
    - Fair selection of promotional rewards
    - Random sampling of transactions for audit

Company Tags: Google, Facebook, Amazon, Apple

See: https://leetcode.com/problems/shuffle-an-array/
"""

import sys
import os
import random
from typing import List

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "..", "..", ".."))

from com.example.tutorial.common.utils.sample_data_loader import load_offers


class Solution:
    """
    Solution class for LeetCode #384.

    LOGIC (Fisher-Yates Shuffle):
        1. Iterate from last element to first
        2. For each position i, pick random index j from 0 to i
        3. Swap elements at i and j
        4. This ensures each permutation is equally likely

    Why It Works:
        For array of n elements:
        - Element at position n-1 has 1/n chance of staying (picking n-1)
        - Element at position n-2 has 1/(n-1) x (n-1)/n = 1/n chance
        - ... each position has exactly 1/n! chance of any permutation

    Time Complexity: O(n)
        Single pass through the array.
        Like shuffling a deck of cards by picking one card at a time
        from the unshuffled portion and placing it in the shuffled portion.

    Space Complexity: O(n)
        For storing the original array.
    """

    def __init__(self, nums: List[int]):
        """
        Initialize with array.

        Args:
            nums: The array to shuffle.
        """
        self.original = nums[:]
        self.current = nums[:]

    def reset(self) -> List[int]:
        """Resets the array to its original configuration."""
        self.current = self.original[:]
        return self.current[:]

    def shuffle(self) -> List[int]:
        """Returns a random shuffling of the array."""
        # Fisher-Yates shuffle.
        for i in range(len(self.current) - 1, 0, -1):
            j = random.randint(0, i)  # Random index from 0 to i.
            # Swap.
            self.current[i], self.current[j] = self.current[j], self.current[i]
        return self.current[:]


class CardDeck:
    """CardDeck class for simulating a deck of playing cards."""

    SUITS = ["♠", "♥", "♦", "♣"]
    RANKS = ["A", "2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K"]

    def __init__(self):
        """Initialize a new deck of cards."""
        self.deck: List[str] = []
        self._initialize_deck()

    def _initialize_deck(self) -> None:
        """Creates a new ordered deck."""
        self.deck = [rank + suit for suit in self.SUITS for rank in self.RANKS]

    def shuffle(self) -> None:
        """Shuffles the deck using Fisher-Yates algorithm."""
        for i in range(len(self.deck) - 1, 0, -1):
            j = random.randint(0, i)
            self.deck[i], self.deck[j] = self.deck[j], self.deck[i]

    def deal(self, n: int) -> List[str]:
        """Deals n cards from the deck."""
        hand = []
        for _ in range(min(n, len(self.deck))):
            hand.append(self.deck.pop(0))
        return hand

    def remaining_cards(self) -> int:
        """Returns number of cards remaining."""
        return len(self.deck)

    def print_top_cards(self, n: int) -> None:
        """Prints the top n cards."""
        print(" ".join(self.deck[:min(n, len(self.deck))]))


def test_shuffle_fairness() -> None:
    """Tests the fairness of the shuffle by counting position frequencies."""
    nums = [1, 2, 3]
    counts = [[0] * 3 for _ in range(3)]  # counts[value][position]
    trials = 30000

    for _ in range(trials):
        # Fisher-Yates shuffle.
        arr = nums[:]
        for i in range(len(arr) - 1, 0, -1):
            j = random.randint(0, i)
            arr[i], arr[j] = arr[j], arr[i]

        # Count positions.
        for i, val in enumerate(arr):
            counts[val - 1][i] += 1

    print(f"Position distribution after {trials} shuffles:")
    print(f"(Expected: ~{trials // 3} each = 33.3%)\n")
    print("        Pos 0   Pos 1   Pos 2")
    for v in range(3):
        print(f"Val {v + 1}:  {counts[v][0]:5d}   {counts[v][1]:5d}   {counts[v][2]:5d}")


def main():
    """Main function to demonstrate the CardShuffler."""
    print("=== CardShuffler: Fisher-Yates Shuffle Demo ===\n")

    # Load credit card offers from sample data
    offers = load_offers()
    print(f"Loaded {len(offers)} offers from sample data.\n")

    # Demonstrate array shuffling
    print("--- Array Shuffle (LeetCode #384) ---\n")
    nums = [1, 2, 3, 4, 5]
    solution = Solution(nums)

    print(f"Original: {nums}")
    print(f"Shuffle 1: {solution.shuffle()}")
    print(f"Shuffle 2: {solution.shuffle()}")
    print(f"Reset: {solution.reset()}")
    print(f"Shuffle 3: {solution.shuffle()}")

    # Demonstrate card deck shuffling
    print("\n--- Card Deck Shuffling ---\n")
    deck = CardDeck()
    print("New deck (first 13 cards):")
    deck.print_top_cards(13)

    deck.shuffle()
    print("\nAfter shuffle (first 13 cards):")
    deck.print_top_cards(13)

    # Deal some cards
    print("\n--- Dealing Cards ---\n")
    print("Dealing 5 cards:")
    hand = deck.deal(5)
    for card in hand:
        print(f"  {card}")

    print(f"\nCards remaining: {deck.remaining_cards()}")

    # Test fairness of shuffle
    print("\n--- Shuffle Fairness Test ---\n")
    test_shuffle_fairness()


if __name__ == "__main__":
    main()

File: src/main/java/com/example/tutorial/dsa/medium/logicgames/CardShuffler.java
package com.example.tutorial.dsa.medium.logicgames;

import com.example.tutorial.common.datamodel.Offer;
import com.example.tutorial.common.utils.SampleDataLoader;
import org.springframework.boot.CommandLineRunner;
import org.springframework.boot.SpringApplication;
import org.springframework.stereotype.Component;

import java.util.*;

/**
 * CardShuffler
 * ----------------------------------
 * <p>This program implements the Fisher-Yates shuffle algorithm for cards.
 * The core problem solved here is Shuffle an Array (LeetCode #384).
 *
 * <p><b>Problem Statement:</b>
 * Given an integer array nums, design an algorithm to randomly shuffle the array.
 * All permutations should be equally likely.
 *
 * <p><b>Real UseCase:</b> In a credit card offers system:
 * <ul>
 *   <li>Randomize offer presentation order</li>
 *   <li>Fair selection of promotional rewards</li>
 *   <li>Random sampling of transactions for audit</li>
 * </ul>
 *
 * <p><b>Company Tags:</b> Google, Facebook, Amazon, Apple
 *
 * @see <a href="https://leetcode.com/problems/shuffle-an-array/">LeetCode 384 - Shuffle an Array</a>
 */
@Component
public class CardShuffler implements CommandLineRunner {
  public static void main(String[] args) {
    SpringApplication.run(CardShuffler.class, args);
  }

  @Override
  public void run(String... args) throws Exception {
    System.out.println("=== CardShuffler: Fisher-Yates Shuffle Demo ===\n");

    // Load credit card offers from sample data
    List<Offer> offers = SampleDataLoader.OFFERS_DTO.get();
    System.out.println("Loaded " + offers.size() + " offers from sample data.\n");

    // Demonstrate array shuffling
    System.out.println("--- Array Shuffle (LeetCode #384) ---\n");
    int[] nums = {1, 2, 3, 4, 5};
    Solution solution = new Solution(nums);

    System.out.println("Original: " + Arrays.toString(nums));
    System.out.println("Shuffle 1: " + Arrays.toString(solution.shuffle()));
    System.out.println("Shuffle 2: " + Arrays.toString(solution.shuffle()));
    System.out.println("Reset: " + Arrays.toString(solution.reset()));
    System.out.println("Shuffle 3: " + Arrays.toString(solution.shuffle()));

    // Demonstrate card deck shuffling
    System.out.println("\n--- Card Deck Shuffling ---\n");
    CardDeck deck = new CardDeck();
    System.out.println("New deck (first 13 cards):");
    deck.printTopCards(13);

    deck.shuffle();
    System.out.println("\nAfter shuffle (first 13 cards):");
    deck.printTopCards(13);

    // Deal some cards
    System.out.println("\n--- Dealing Cards ---\n");
    System.out.println("Dealing 5 cards:");
    List<String> hand = deck.deal(5);
    for (String card : hand) {
      System.out.println("  " + card);
    }

    System.out.println("\nCards remaining: " + deck.remainingCards());

    // Test fairness of shuffle
    System.out.println("\n--- Shuffle Fairness Test ---\n");
    testShuffleFairness();
  }

  /**
   * Solution class for LeetCode #384.
   *
   * <p><b>LOGIC (Fisher-Yates Shuffle):</b>
   * <ol>
   *   <li>Iterate from last element to first</li>
   *   <li>For each position i, pick random index j from 0 to i</li>
   *   <li>Swap elements at i and j</li>
   *   <li>This ensures each permutation is equally likely</li>
   * </ol>
   *
   * <p><b>Why It Works:</b>
   * <pre>
   * For array of n elements:
   * - Element at position n-1 has 1/n chance of staying (picking n-1)
   * - Element at position n-2 has 1/(n-1) × (n-1)/n = 1/n chance
   * - ... each position has exactly 1/n! chance of any permutation
   * </pre>
   *
   * <p><b>Time Complexity: O(n)</b>
   * <br>Single pass through the array.
   * <br><i>Like shuffling a deck of cards by picking one card at a time
   * from the unshuffled portion and placing it in the shuffled portion.</i>
   *
   * <p><b>Space Complexity: O(n)</b>
   * <br>For storing the original array.
   */
  public static class Solution {
    private final int[] original;
    private final int[] current;
    private final Random random;

    public Solution(int[] nums) {
      this.original = nums.clone();
      this.current = nums.clone();
      this.random = new Random();
    }

    /**
     * Resets the array to its original configuration.
     */
    public int[] reset() {
      System.arraycopy(original, 0, current, 0, original.length);
      return current.clone();
    }

    /**
     * Returns a random shuffling of the array.
     */
    public int[] shuffle() {
      // Fisher-Yates shuffle.
      for (int i = current.length - 1; i > 0; i--) {
        int j = random.nextInt(i + 1);  // Random index from 0 to i.
        // Swap.
        int temp = current[i];
        current[i] = current[j];
        current[j] = temp;
      }
      return current.clone();
    }
  }

  /**
   * CardDeck class for simulating a deck of playing cards.
   */
  public static class CardDeck {
    private static final String[] SUITS = {"♠", "♥", "♦", "♣"};
    private static final String[] RANKS = {
        "A", "2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K"
    };

    private final List<String> deck;
    private final Random random;

    public CardDeck() {
      this.deck = new ArrayList<>();
      this.random = new Random();
      initializeDeck();
    }

    private void initializeDeck() {
      deck.clear();
      for (String suit : SUITS) {
        for (String rank : RANKS) {
          deck.add(rank + suit);
        }
      }
    }

    /**
     * Shuffles the deck using Fisher-Yates algorithm.
     */
    public void shuffle() {
      for (int i = deck.size() - 1; i > 0; i--) {
        int j = random.nextInt(i + 1);
        Collections.swap(deck, i, j);
      }
    }

    /**
     * Deals n cards from the deck.
     */
    public List<String> deal(int n) {
      List<String> hand = new ArrayList<>();
      for (int i = 0; i < n && !deck.isEmpty(); i++) {
        hand.add(deck.remove(0));
      }
      return hand;
    }

    public int remainingCards() {
      return deck.size();
    }

    public void printTopCards(int n) {
      for (int i = 0; i < Math.min(n, deck.size()); i++) {
        System.out.print(deck.get(i) + " ");
      }
      System.out.println();
    }
  }

  /**
   * Tests the fairness of the shuffle by counting position frequencies.
   */
  private static void testShuffleFairness() {
    int[] nums = {1, 2, 3};
    int[][] counts = new int[3][3];  // counts[value][position]
    int trials = 30000;

    Random random = new Random();

    for (int t = 0; t < trials; t++) {
      // Fisher-Yates shuffle.
      int[] arr = nums.clone();
      for (int i = arr.length - 1; i > 0; i--) {
        int j = random.nextInt(i + 1);
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
      }

      // Count positions.
      for (int i = 0; i < arr.length; i++) {
        counts[arr[i] - 1][i]++;
      }
    }

    System.out.println("Position distribution after " + trials + " shuffles:");
    System.out.println("(Expected: ~" + (trials / 3) + " each = 33.3%)\n");
    System.out.println("        Pos 0   Pos 1   Pos 2");
    for (int v = 0; v < 3; v++) {
      System.out.printf("Val %d:  %5d   %5d   %5d%n",
          v + 1, counts[v][0], counts[v][1], counts[v][2]);
    }
  }
}
